Add crypto.x509.parse_and_verify_certificates

jalseth opened the pull request:
This allows users to supply a certificate chain and verify that the leaf
node has a path back to the trusted root.

Signed-off-by: James Alseth <james@jalseth.me>

---

I haven't added tests or updated the docs yet, as I'd appreciate feedback on the code change before I work on those to avoid rework.

Implements #3601

srenatus commented on topdown/crypto.go:
```diff
@@ -160,12 +164,68 @@ func builtinCryptoSha256(a ast.Value) (ast.Value, error) {
 	RegisterFunctionalBuiltin1(ast.CryptoX509ParseCertificateRequest.Name, builtinCryptoX509ParseCertificateRequest)
 }
 
+func getX509CertsFromBytes(bytes []byte) ([]*x509.Certificate, error) {
+	// if the input is not a PEM string, attempt to decode b64
+	if str := string(bytes); !strings.HasPrefix(str, "-----BEGIN CERTIFICATE-----") {
+		var err error
+		bytes, err = base64.StdEncoding.DecodeString(str)
+		if err != nil {
+			return nil, err
+		}
+	}
```
[nit] I don't think it matters for performance at all, but we're only feeding this `[]byte(someString)` -- we might as well go with
```suggestion
func getX509CertsFromBytes(certs string) ([]*x509.Certificate, error) {
	var decodedCerts []byte
	// if the input is not a PEM string, attempt to decode b64
	if strings.HasPrefix(certs, "-----BEGIN CERTIFICATE-----") {
		decodedCerts = []byte(certs)
	} else {
		var err error
		decodedCerts, err = base64.StdEncoding.DecodeString(str)
		if err != nil {
			return nil, err
		}
	}
```
Well, not a big difference. Just do as you see fit 😉 

Let's try to avoid calling variables by their type (`var bytes []byte`) but use something indicating their contents. 😃

jalseth replied:
Makes sense, I'll update this as well as the function name.

srenatus commented on topdown/crypto.go:
```diff
@@ -160,12 +164,68 @@ func builtinCryptoSha256(a ast.Value) (ast.Value, error) {
+
+	// attempt to decode input as PEM data
+	p, rest := pem.Decode(bytes)
+	if p != nil && p.Type != blockTypeCertificate {
```
❓ We're having three `if p != nil` here, would it perhaps be preferable to go with `if p == nil { /* return something */ }` and go ahead with `p` if not? (The check on line 199 is different, of course.)

jalseth replied:
This actually isn't related to this change. This was the exiting body of `parse_certificates` which I just moved out into its own function so it could be reused by `parse_and_verify_certificates`. I didn't look into the code at all since it was here previously and appears to work. That said, now that I've looked I agree this could be improved.

srenatus commented on topdown/crypto.go:
```diff
@@ -160,12 +164,68 @@ func builtinCryptoSha256(a ast.Value) (ast.Value, error) {
+	return x509.ParseCertificates(bytes)
+}
+
+func marshalX509Certs(certs []*x509.Certificate) (ast.Value, error) {
+	bs, err := json.Marshal(certs)
+	if err != nil {
+		return nil, err
+	}
+
+	var x interface{}
+
+	if err := util.UnmarshalJSON(bs, &x); err != nil {
+		return nil, err
+	}
+
+	return ast.InterfaceToValue(x)
+}
```
❓ Do we still need this, given https://github.com/open-policy-agent/opa/pull/3586?

srenatus commented on topdown/crypto.go:
```diff
@@ -32,67 +32,71 @@ const (
-		bytes = p.Bytes
+
+	if len(certs) < 2 {
+		return fmt.Errorf("must supply at least two certificates to verify the chain")
```
❓ I suppose this would be an _operand error_? We've got a type for that, I believe.

jalseth replied:
Thanks, I wasn't aware of that.

srenatus commented on topdown/crypto.go:
```diff
@@ -32,67 +32,71 @@ const (
+}
+
+func builtinCryptoX509ParseAndVerifyCertificates(
+	bctx BuiltinContext, args []*ast.Term, iter func(*ast.Term) error) error {
```
```suggestion
	_ BuiltinContext, args []*ast.Term, iter func(*ast.Term) error) error {
```

srenatus reviewed (commented):
🎉 **Thank you for your contribution.** Some really minor nitpicks inline, this looks good! 👍

jalseth commented:
@srenatus This should be set now. I addressed your feedback, added tests and docs, rebased on the current main branch, and cleaned up the commit history.

srenatus commented on docs/content/policy-reference.md:
```diff
@@ -793,6 +793,7 @@ Note that the opa executable will need access to the timezone files in the envir
 | Built-in | Description | Wasm Support |
 | ------- |-------------|---------------|
 | <span class="opa-keep-it-together">``output := crypto.x509.parse_certificates(certs)``</span> | ``certs`` is base64 encoded DER or PEM data containing one or more certificates or a PEM string of one or more certificates. ``output`` is an array of X.509 certificates represented as JSON objects. | ``SDK-dependent`` |
+| <span class="opa-keep-it-together">``output := crypto.x509.parse_and_verify_certificates(certs)``</span> | ``certs`` is base64 encoded DER or PEM data containing two or more certificates where the first is a root CA, the last is a leaf certificate, and all others are intermediate CAs. ``output`` is of the form ``[valid, certs]``. If the input certificate chain could be verified then ``valid`` i ...
```
🤔 Is this in line with what comparable builtins do? I'd have expected `[valid, certs]` or `undefined` (i.e. returning some builtin-error type that would end up in an evaluation to `undefined` if not using strict builtin errors).

jalseth replied:
Are you referring to the choice of returning an empty array if `valid` is false? If so, then yes this is in line with what JWT `decode_verify` returns. If someone is explicitly checking the cert chain, I think it makes sense to not return the certs if the verification fails to make sure the non-verified certs aren't accidentally trusted in the policy.

srenatus replied:
I'm thinking about this more broadly -- why return anything if `valid` is false? We could return the `certs` array if it's valid, and not return anything at all if not.

But I've been scanning the implementation of other ` io.jwt.*` functions (like `io.jwt.verify_rs256`), and they do return true or false (not true or undefined). So, yeah, let's keep this as you've implemented it, it should be the choice more consistent with the rest 👍

srenatus reviewed (commented):
😃 This looks good, thanks for bearing with me!

One question re: semantics of validation failures 👇

srenatus reviewed (approved):
LGTM

srenatus commented:
@jalseth I was about to merge this when updating the branch and re-running the tests made them fail.

Looking into it a bit, it seems it's a problem with the testdata certs,

```
$ go test ./topdown -v -run TestX509ParseAndVerify
=== RUN   TestX509ParseAndVerify
=== RUN   TestX509ParseAndVerify/TestFullChainPEM
    crypto_test.go:56: x509 verification failed when it was expected to succeed: x509: certificate has expired or is not yet valid: current time 2021-07-05T10:07:40+02:00 is after 2021-07-02T18:20:34Z:
=== RUN   TestX509ParseAndVerify/TestFullChainBase64
    crypto_test.go:70: x509 verification failed when it was expected to succeed: x509: certificate has expired or is not yet valid: current time 2021-07-05T10:07:40+02:00 is after 2021-07-02T18:20:34Z:
=== RUN   TestX509ParseAndVerify/TestWrongOrder
=== RUN   TestX509ParseAndVerify/TestMissingIntermediate
=== RUN   TestX509ParseAndVerify/TestTooFewCerts
--- FAIL: TestX509ParseAndVerify (0.00s)
    --- FAIL: TestX509ParseAndVerify/TestFullChainPEM (0.00s)
    --- FAIL: TestX509ParseAndVerify/TestFullChainBase64 (0.00s)
    --- PASS: TestX509ParseAndVerify/TestWrongOrder (0.00s)
    --- PASS: TestX509ParseAndVerify/TestMissingIntermediate (0.00s)
    --- PASS: TestX509ParseAndVerify/TestTooFewCerts (0.00s)
FAIL
FAIL    github.com/open-policy-agent/opa/topdown        0.204s
FAIL
```

Would you mind having another look here? Thank you!

jalseth commented:
@srenatus Sorry about that, I forgot the default lifetime of certificates created with `step` is only one day, so the leaf had expired. I've generated a new leaf and the chain is now good for 10 years. Let me know if you think this needs to be longer.

srenatus commented:
I'll check the race tests.

srenatus commented:
@jalseth something's flaky there. Re-running the jobs made it pass on Github. I've been running this locally in a loop for a while now (without failures), but I might try again with more limited resources (mimicking a Github Actions runner) soon... As you indicated, this is unrelated to this PR. Thanks again! 🚀
